sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiverClient.cs

```diff
@@ -483,197 +494,696 @@ public virtual async Task<ServiceBusMessage> PeekAsync(CancellationToken cancell
+                Error error = null;
+                foreach (Outcome item in outcomes)
+                {
+                    var disposedOutcome = item.DescriptorCode == Rejected.Code && ((error = ((Rejected)item).Error) != null) ? item : null;
+                    if (disposedOutcome != null)
+                    {
+                        if (error.Condition.Equals(AmqpErrorCode.NotFound))
+                        {
+                            if (IsSessionReceiver)
+                            {
+                                //  throw new SessionLockLostException(Resources.SessionLockExpiredOnMessageSession);
+                            }
+
+                            //   throw new MessageLockLostException(Resources.MessageLockLost);
+                        }
+
+                        // throw error.ToMessagingContractException();
+                    }
+                }
+            }
+            catch (Exception exception)
+            {
+                if (exception is OperationCanceledException &&
+                    receiveLink != null && receiveLink.State != AmqpObjectState.Opened)
+                {
+                    // The link state is lost, We need to return a non-retriable error.
+                    // MessagingEventSource.Log.LinkStateLost(this.ClientId, receiveLink.Name, receiveLink.State, this.isSessionReceiver, exception);
+                    if (IsSessionReceiver)
+                    {
+                        //  throw new SessionLockLostException(Resources.SessionLockExpiredOnMessageSession);
+                    }
+
+                    // throw new MessageLockLostException(Resources.MessageLockLost);
+                }
+
+                // throw AmqpExceptionHelper.GetClientException(exception);
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// Updates the disposition status of deferred messages.
+        /// </summary>
+        ///
+        /// <param name="lockTokens">Message lock tokens to update disposition status.</param>
+        /// <param name="timeout"></param>
+        /// <param name="dispositionStatus"></param>
+        /// <param name="propertiesToModify"></param>
+        /// <param name="deadLetterReason"></param>
+        /// <param name="deadLetterDescription"></param>
+        internal async Task DisposeMessageRequestResponseAsync(
+            Guid[] lockTokens,
+            TimeSpan timeout,
+            DispositionStatus dispositionStatus,
+            IDictionary<string, object> propertiesToModify = null,
+            string deadLetterReason = null,
+            string deadLetterDescription = null)
+        {
+            try
+            {
+                // Create an AmqpRequest Message to update disposition
+                var amqpRequestMessage = AmqpRequestMessage.CreateRequest(ManagementConstants.Operations.UpdateDispositionOperation, timeout, null);
```

Comment:
Can we move this into AMQP layer? The goal is to have all transport specific stuff be in the transport layer so that in theory we could support multiple transport types. So any thing with AMQP in the name should ideally not be at this layer.

Reply:
Moved implementation to the Amqp layer.